flips/20220111-execution-effort.md

```diff
@@ -0,0 +1,555 @@
+| UpdateAccountContractCode  | function calls                           | Updates an account's contract                                                                               |
+| ValueExists                | function calls                           | Checks if a certain register exists                                                                         |
+| ValueDecoded               | function calls                           | Cadence callback after a value was decoded into a cadence type                                              |
+| ValueEncoded               | function calls                           | Cadence callback after a cadence type was encoded into a cadence json                                       |
+
+### Data analysis
+
+The large amount of different weights chosen was intentional. This allows for better insight into which weights should actually be used.
+
+The Following chapters describe how the linear model fitting was done due to a significant amount of outliers in the data and how certain weights were eliminated, because they were either redundant (correlated with other weights), insignificant (had small or no impact on the correlation), or had a lot of noise.
+
+A note on the `time to execute [ms]` vs `Execution effort` graphs. The closer the data is to the red dashed line which represents `time to execute = Execution effort` the better the fit of the model. Data that is to the right side of the graph, represents transactions that took longer than expected, and were thus charged less than they should be (but some variance and outliers are expected here). Data to the left side of the graph represents transactions where the execution did not take as long as the model predicted, and as a consequence those transactions will be charged more then they should be.
+
+#### Weight correlation
+
+Before using the data in a linear model, its important to note that the rank of the matrix of all weight data <!-- $M$ --> <img style="transform: translateY(0.1em); background: white;" src="./20220111-execution-effort/eq/SmblEwKco6.svg"> is smaller than the number of weights. This means that some wights are correlated to others. These are asy to spot by checking the correlation matrix where only the entries with 1 or -1 are highlighted.
+
+![Correlation1](./20220111-execution-effort/correlation1.png)
+
+The pairs that are correlated are:
+
+- `ProgramChecked` / `ProgramParsed`
+- `GetCode` / `GetAccountContractCode`
+- `ValueExists` / `CreateAccount`
+
+These pairs of weights were cross checked with the implementation to confirm that they are always called together.
+
+The columns with weights `ProgramChecked`, `GetCode` and `ValueExists` were removed from the weight data matrix <!-- $M$ --> <img style="transform: translateY(0.1em); background: white;" src="./20220111-execution-effort/eq/SmblEwKco6.svg">.
+
+#### Outliers and robust linear model fitting
```

Comment:
😨 I am a bit worried here about the use if robust linear model fitting. As Wolfram explained, robust linear model fitting introduces an axiomatic bias into the model. There is a strong risk here of over-fitting to the training data (benchmark transactions),  so its predictive performance on the training set is better, but potentially worse for unknown transactions. Meaning, the model-generated "execution effort" would not be a good predictor for real-world effort.

Reply:
There is a property of the data I tried to include in the model, the "robust fit" + custom weighting function for outliers Is what I used to try to do this.

The property I had in mind (and is also visible in the data) is that for a specific transaction on a specific machine there exists a `t_0` such that the execution time for this transaction on this machine (across multiple runs) is `t=t_min + delta t` where `delta t >= 0`. In other words the machine cannot run the transaction faster than `t_0` but it can (and will) run it slower. (The error of execution time is more positive than negative).